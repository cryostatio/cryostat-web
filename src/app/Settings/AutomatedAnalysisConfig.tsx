/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { AutomatedAnalysisConfigForm } from '@app/Dashboard/AutomatedAnalysis/AutomatedAnalysisConfigForm';
import { ServiceContext } from '@app/Shared/Services/Services';
import { AutomatedAnalysisRecordingConfig } from '@app/Shared/Services/Settings.service';
import { TargetView } from '@app/TargetView/TargetView';
import {
  DescriptionList,
  DescriptionListDescription,
  DescriptionListGroup,
  DescriptionListTerm,
  ExpandableSection,
  Stack,
  StackItem,
  Title,
} from '@patternfly/react-core';
import * as React from 'react';
import { UserSetting } from './Settings';

export const defaultAutomatedAnalysisRecordingConfig: AutomatedAnalysisRecordingConfig = {
  templates: 'template=Continuous,type=TARGET',
  maxSize: 2048,
  maxAge: 0,
};

const Component = () => {
  const context = React.useContext(ServiceContext);
  const [config, setConfig] = React.useState<AutomatedAnalysisRecordingConfig>(defaultAutomatedAnalysisRecordingConfig);
  const [expanded, setExpanded] = React.useState(false);

  React.useEffect(() => {
    const c = context.settings.automatedAnalysisRecordingConfig();
    setConfig(c);
  }, []);

  const onSave = React.useCallback(() => {
    const newConfig = context.settings.automatedAnalysisRecordingConfig();
    setConfig(newConfig);
  }, [context.settings, context.settings.automatedAnalysisRecordingConfig, setConfig]);

  return (
    <TargetView pageTitle={''} compactSelect>
      <Stack hasGutter>
        <StackItem>
          <Title headingLevel="h2" size="lg">
            Current configuration
          </Title>
        </StackItem>
        <StackItem>
          <DescriptionList columnModifier={{ lg: '3Col' }}>
            <DescriptionListGroup>
              <DescriptionListTerm>Template</DescriptionListTerm>
              <DescriptionListDescription>{config.templates}</DescriptionListDescription>
            </DescriptionListGroup>
            <DescriptionListGroup>
              <DescriptionListTerm>Max Size (B)</DescriptionListTerm>
              <DescriptionListDescription>{config.maxSize}</DescriptionListDescription>
            </DescriptionListGroup>
            <DescriptionListGroup>
              <DescriptionListTerm>Max Age (s)</DescriptionListTerm>
              <DescriptionListDescription>{config.maxAge}</DescriptionListDescription>
            </DescriptionListGroup>
          </DescriptionList>
        </StackItem>
        <ExpandableSection
          toggleText={expanded ? 'Show less' : 'Edit configuration'}
          onToggle={setExpanded}
          isExpanded={expanded}
        >
          <AutomatedAnalysisConfigForm onSave={onSave} isSettingsForm={true} />
        </ExpandableSection>
      </Stack>
    </TargetView>
  );
};

export const AutomatedAnalysisConfig: UserSetting = {
  title: 'Automated Analysis Recording Configuration',
  description:
    'Set the recording configuration for automated analysis recordings. You may want smaller or larger values for max-age and max-size depending on how recent you want events to be recorded from the analysis.',
  content: Component,
};
