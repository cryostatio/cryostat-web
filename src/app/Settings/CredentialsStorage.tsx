/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import * as React from 'react';
import { Link } from 'react-router-dom';
import { Select, SelectOption, SelectVariant, Text } from '@patternfly/react-core';
import { UserSetting } from './Settings';
import { getFromLocalStorage, saveToLocalStorage } from '@app/utils/LocalStorage';

export interface Location {
  key: string;
  description: string;
}

export class Locations {
  static readonly BROWSER_SESSION: Location = {
    key: 'Session (Browser Memory)',
    description:
      'Keep credentials in browser memory for the current session only. When you close this browser tab the credentials will be forgotten.',
  };
  static readonly BACKEND: Location = {
    key: 'Backend',
    description:
      'Keep credentials in encrypted Cryostat backend storage. These credentials will be available to other users and will be used for Automated Rules.',
  };
}

const locations = [Locations.BROWSER_SESSION, Locations.BACKEND];

const getLocation = (key: string): Location => {
  for (let l of locations) {
    if (l.key === key) {
      return l;
    }
  }
  return Locations.BACKEND;
};

const Component = () => {
  const [isExpanded, setExpanded] = React.useState(false);
  const [selection, setSelection] = React.useState(Locations.BACKEND.key);

  const handleSelect = React.useCallback(
    (_, selection) => {
      let location = getLocation(selection);
      setSelection(location.key);
      setExpanded(false);
      saveToLocalStorage('JMX_CREDENTIAL_LOCATION', selection);
    },
    [getLocation, setSelection, setExpanded, saveToLocalStorage]
  );

  React.useEffect(() => {
    handleSelect(undefined, getFromLocalStorage('JMX_CREDENTIAL_LOCATION', Locations.BACKEND.key));
  }, [handleSelect, getFromLocalStorage]);

  return (
    <>
      <Select
        variant={SelectVariant.single}
        onToggle={setExpanded}
        onSelect={handleSelect}
        isOpen={isExpanded}
        selections={selection}
      >
        {locations.map((location) => (
          <SelectOption key={location.key} value={location.key} description={location.description} />
        ))}
      </Select>
    </>
  );
};

export const CredentialsStorage: UserSetting = {
  title: 'JMX Credentials Storage',
  description: (
    <Text>
      When you attempt to connect to a target application which requires authentication, you will see a prompt for
      credentials to present to the application and complete the connection. You can choose where to persist these
      credentials. Any credentials added through the <Link to="/security">Security</Link> panel will always be stored in
      Cryostat backend encrypted storage.
    </Text>
  ),
  content: Component,
};
