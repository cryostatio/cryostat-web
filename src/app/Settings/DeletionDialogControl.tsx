/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import * as React from 'react';
import { Divider, ExpandableSection, Switch, Stack, StackItem } from '@patternfly/react-core';
import { ServiceContext } from '@app/Shared/Services/Services';
import { UserSetting } from './Settings';
import { DeleteWarningType, getFromWarningMap } from '@app/Modal/DeleteWarningUtils';

const Component = () => {
  const context = React.useContext(ServiceContext);
  const [state, setState] = React.useState(context.settings.deletionDialogsEnabled());
  const [expanded, setExpanded] = React.useState(false);

  const handleCheckboxChange = React.useCallback(
    (checked, element) => {
      state.set(DeleteWarningType[element.target.id], checked);
      context.settings.setDeletionDialogsEnabled(state);
      setState(new Map(state));
    },
    [state, setState, context.settings]
  );

  const handleCheckAll = React.useCallback(
    (checked) => {
      const newState = new Map();
      Array.from(state.entries()).forEach((v) => newState.set(v[0], checked));
      context.settings.setDeletionDialogsEnabled(newState);
      setState(newState);
    },
    [state, setState]
  );

  const allChecked = React.useMemo(() => {
    return Array.from(state.entries())
      .map((e) => e[1])
      .reduce((a, b) => a && b);
  }, [state]);

  const switches = React.useMemo(() => {
    return Array.from(state.entries(), ([key, value]) => (
      <StackItem key={key}>
        <Switch
          id={key}
          label={getFromWarningMap(key)?.label || key.toString()}
          isChecked={value}
          onChange={handleCheckboxChange}
        />
      </StackItem>
    ));
  }, [state]);

  return (
    <>
      <Stack hasGutter>
        <StackItem key="all-deletion-warnings">
          <Switch
            id="all-deletion-warnings"
            label="All Deletion Warnings"
            isChecked={allChecked}
            onChange={handleCheckAll}
          />
        </StackItem>
        <Divider />
        <ExpandableSection
          toggleText={expanded ? 'Show less' : 'Show more'}
          onToggle={setExpanded}
          isExpanded={expanded}
        >
          {switches}
        </ExpandableSection>
      </Stack>
    </>
  );
};

export const DeletionDialogControl: UserSetting = {
  title: 'Show Deletion Dialogs',
  description: 'Enable or disable deletion dialogs by deletion type.',
  content: Component,
};
