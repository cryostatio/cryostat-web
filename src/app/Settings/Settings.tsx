/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BreadcrumbPage } from '@app/BreadcrumbPage/BreadcrumbPage';
import { FeatureFlag } from '@app/Shared/Components/FeatureFlag';
import { FeatureLevel } from '@app/Shared/Services/service.types';
import { cleanDataId, getActiveTab, hashCode, switchTab } from '@app/utils/utils';
import {
  Card,
  Form,
  FormGroup,
  FormHelperText,
  HelperText,
  HelperTextItem,
  Label,
  Sidebar,
  SidebarContent,
  SidebarPanel,
  Tab,
  TabProps,
  Tabs,
  TabTitleText,
  Title,
} from '@patternfly/react-core';
import * as React from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';
import { AutomatedAnalysis } from './Config/AutomatedAnalysis';
import { AutoRefresh } from './Config/AutoRefresh';
import { ChartCards } from './Config/ChartCards';
import { DatetimeControl } from './Config/DatetimeControl';
import { DeletionDialogControl } from './Config/DeletionDialogControl';
import { FeatureLevels } from './Config/FeatureLevels';
import { Language } from './Config/Language';
import { NotificationControl } from './Config/NotificationControl';
import { Theme } from './Config/Theme';
import { WebSocketDebounce } from './Config/WebSocketDebounce';
import { SettingGroup, SettingTab, _TransformedUserSetting } from './types';
import { paramAsTab, tabAsParam, getGroupFeatureLevel } from './utils';

export interface SettingsProps {}

export const Settings: React.FC<SettingsProps> = (_) => {
  const { t } = useTranslation();

  const settings = React.useMemo(
    () =>
      [
        NotificationControl,
        AutomatedAnalysis,
        ChartCards,
        DeletionDialogControl,
        WebSocketDebounce,
        AutoRefresh,
        FeatureLevels,
        Language,
        DatetimeControl,
        Theme,
      ].map(
        (c) =>
          ({
            title: t(c.titleKey) || '',
            description:
              typeof c.descConstruct === 'string' ? (
                t(c.descConstruct)
              ) : (
                // Use children prop to avoid i18n parses body as key
                /* eslint react/no-children-prop: 0 */
                <Trans i18nKey={c.descConstruct.key} children={c.descConstruct.parts} />
              ),
            element: React.createElement(c.content, null),
            category: c.category,
            disabled: c.disabled,
            orderInGroup: c.orderInGroup || -1,
            featureLevel: c.featureLevel || FeatureLevel.PRODUCTION,
          }) as _TransformedUserSetting,
      ),
    [t],
  );

  const navigate = useNavigate();
  const { search, pathname } = useLocation();

  const activeTab = React.useMemo(() => {
    return paramAsTab(
      getActiveTab(
        search,
        'tab',
        Object.values(SettingTab).map((v) => tabAsParam(v)),
        tabAsParam(SettingTab.GENERAL),
      ),
    );
  }, [search]);

  const onTabSelect = React.useCallback(
    (_: React.MouseEvent, key: string | number) =>
      switchTab(navigate, pathname, search, { tabKey: 'tab', tabValue: `${tabAsParam(key as SettingTab)}` }),
    [navigate, pathname, search],
  );

  const settingGroups = React.useMemo(() => {
    return Object.values(SettingTab).map((cat) => {
      const panels = settings.filter((s) => s.category === cat).sort((a, b) => b.orderInGroup - a.orderInGroup);
      return {
        groupLabel: t(cat),
        groupKey: cat,
        settings: panels,
        featureLevel: getGroupFeatureLevel(panels),
      };
    }) as SettingGroup[];
  }, [settings, t]);

  return (
    <>
      <BreadcrumbPage pageTitle="Settings">
        <Card isFullHeight>
          <Sidebar tabIndex={0}>
            <SidebarPanel>
              <Tabs
                isVertical
                isExpanded
                aria-label="Select setting category"
                activeKey={activeTab}
                onSelect={onTabSelect}
              >
                {settingGroups.map((grp) => (
                  <WrappedTab
                    key={`${grp.groupLabel}-tab`}
                    eventKey={grp.groupKey}
                    title={<TabTitleText>{grp.groupLabel}</TabTitleText>}
                    featureLevelConfig={{
                      level: grp.featureLevel,
                    }}
                  />
                ))}
              </Tabs>
            </SidebarPanel>
            <SidebarContent>
              {settingGroups
                .filter((grp) => grp.groupKey === activeTab)
                .map((grp) => (
                  <Form key={`${grp.groupKey}-setting`} className="settings__content">
                    {grp.settings.map((s, index) => (
                      <FeatureFlag level={s.featureLevel} key={`${grp.groupLabel}-${s.title}-${index}-flag`}>
                        <FormGroup
                          label={
                            <Title headingLevel={'h2'} size={'lg'}>
                              {s.title}
                              {s.featureLevel !== FeatureLevel.PRODUCTION && (
                                <Label
                                  isCompact
                                  style={{
                                    marginLeft: '1ch',
                                    textTransform: 'capitalize',
                                  }}
                                  color={s.featureLevel === FeatureLevel.BETA ? 'green' : 'red'}
                                >
                                  {FeatureLevel[s.featureLevel].toLowerCase()}
                                </Label>
                              )}
                            </Title>
                          }
                          key={`${grp.groupLabel}-${s.title}-${index}`}
                        >
                          <FormHelperText>
                            <HelperText>
                              <HelperTextItem>{s.description}</HelperTextItem>
                            </HelperText>
                          </FormHelperText>
                          {s.element}
                        </FormGroup>
                      </FeatureFlag>
                    ))}
                  </Form>
                ))}
            </SidebarContent>
          </Sidebar>
        </Card>
        <>
          {
            // Need this fragment to correct bottom margin.
          }
        </>
      </BreadcrumbPage>
    </>
  );
};

interface SettingTabProps extends TabProps {
  featureLevelConfig: {
    level: FeatureLevel;
    strict?: boolean;
  };
}

// Workaround to the Tabs component requiring children to be React.FC<TabProps>
const WrappedTab: React.FC<SettingTabProps> = ({ featureLevelConfig, eventKey, title, children }) => {
  const { t } = useTranslation();

  return (
    <FeatureFlag level={featureLevelConfig.level} strict={featureLevelConfig.strict}>
      <Tab
        eventKey={eventKey}
        title={title}
        id={`${hashCode(`${eventKey}`)}`}
        data-quickstart-id={cleanDataId(`settings-${t(eventKey.toString())}-tab`)}
      >
        {children}
      </Tab>
    </FeatureFlag>
  );
};

export default Settings;
