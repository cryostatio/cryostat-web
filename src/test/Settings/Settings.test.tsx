/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import '@testing-library/jest-dom';
import renderer, { act } from 'react-test-renderer';
import { cleanup } from '@testing-library/react';
import { Text } from '@patternfly/react-core';
import { Settings } from '@app/Settings/Settings';

// jest.spyOn(defaultServices.target, 'target').mockReturnValue(of(mockTarget));

jest.mock('@app/Settings/NotificationControl', () => ({
  NotificationControl: {
    title: 'Notification Control Title',
    description: 'Notification Control Description',
    content: () => <Text>Notification Control Component</Text>,
  },
}));

jest.mock('@app/Settings/CredentialsStorage', () => ({
  CredentialsStorage: {
    title: 'Credentials Storage Title',
    description: 'Credentials Storage Description',
    content: () => <Text>Credentials Storage Component</Text>,
  },
}));

jest.mock('@app/Settings/DeletionDialogControl', () => ({
  DeletionDialogControl: {
    title: 'Deletion Dialog Control Title',
    description: 'Deletion Dialog Control Description',
    content: () => <Text>Deletion Dialog Control Component</Text>,
  },
}));

jest.mock('@app/Settings/WebSocketDebounce', () => ({
  WebSocketDebounce: {
    title: 'WebSocket Debounce Title',
    description: 'WebSocket Debounce Description',
    content: () => <Text>WebSocket Debounce Component</Text>,
  },
}));

jest.mock('@app/Settings/AutoRefresh', () => ({
  AutoRefresh: {
    title: 'AutoRefresh Title',
    description: 'AutoRefresh Description',
    content: () => <Text>AutoRefresh Component</Text>,
  },
}));

describe('<Settings/>', () => {
  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(<Settings />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
