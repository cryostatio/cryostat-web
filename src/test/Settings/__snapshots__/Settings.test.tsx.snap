// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings/> renders correctly 1`] = `
<div
  className="pf-v5-c-page__main-group"
>
  <section
    className="pf-v5-c-page__main-section"
  >
    <nav
      aria-label="Breadcrumb"
      className="pf-v5-c-breadcrumb"
      data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
      data-ouia-component-type="PF5/Breadcrumb"
      data-ouia-safe={true}
    >
      <ol
        className="pf-v5-c-breadcrumb__list"
        role="list"
      >
        <li
          className="pf-v5-c-breadcrumb__item"
        >
          <h1
            className="pf-v5-c-breadcrumb__heading"
          >
            Settings
          </h1>
        </li>
      </ol>
    </nav>
    <div
      className="pf-v5-l-stack pf-m-gutter"
    >
      <div
        className="pf-v5-l-stack__item pf-m-fill"
      >
        <div
          className="pf-v5-c-card pf-m-full-height"
          data-ouia-component-id="OUIA-Generated-Card-1"
          data-ouia-component-type="PF5/Card"
          data-ouia-safe={true}
          id=""
        >
          <div
            className="pf-v5-c-sidebar"
            tabIndex={0}
          >
            <div
              className="pf-v5-c-sidebar__main"
            >
              <div
                className="pf-v5-c-sidebar__panel"
              >
                <div
                  aria-label="Select setting category"
                  className="pf-v5-c-tabs pf-m-vertical"
                  data-ouia-component-id="OUIA-Generated-Tabs-1"
                  data-ouia-component-type="PF5/Tabs"
                  data-ouia-safe={true}
                  onSelect={[Function]}
                >
                  <ul
                    className="pf-v5-c-tabs__list"
                    onScroll={[Function]}
                    role="tablist"
                  >
                    <li
                      className="pf-v5-c-tabs__item pf-m-current"
                      role="presentation"
                    >
                      <button
                        aria-controls="pf-tab-section-SETTINGS.CATEGORIES.GENERAL--445403487"
                        aria-selected={true}
                        className="pf-v5-c-tabs__link"
                        data-ouia-component-type="PF5/TabButton"
                        data-ouia-safe={true}
                        data-quickstart-id="settings-general-tab"
                        id="pf-tab-SETTINGS.CATEGORIES.GENERAL--445403487"
                        onClick={[Function]}
                        role="tab"
                        type="button"
                      >
                        <span
                          className="pf-v5-c-tabs__item-text"
                        >
                          General
                        </span>
                      </button>
                    </li>
                    <li
                      className="pf-v5-c-tabs__item"
                      role="presentation"
                    >
                      <button
                        aria-controls="pf-tab-section-SETTINGS.CATEGORIES.CONNECTIVITY--1507428482"
                        aria-selected={false}
                        className="pf-v5-c-tabs__link"
                        data-ouia-component-type="PF5/TabButton"
                        data-ouia-safe={true}
                        data-quickstart-id="settings-connectivity-tab"
                        id="pf-tab-SETTINGS.CATEGORIES.CONNECTIVITY--1507428482"
                        onClick={[Function]}
                        role="tab"
                        type="button"
                      >
                        <span
                          className="pf-v5-c-tabs__item-text"
                        >
                          Connectivity
                        </span>
                      </button>
                    </li>
                    <li
                      className="pf-v5-c-tabs__item"
                      role="presentation"
                    >
                      <button
                        aria-controls="pf-tab-section-SETTINGS.CATEGORIES.NOTIFICATION_MESSAGE-719517530"
                        aria-selected={false}
                        className="pf-v5-c-tabs__link"
                        data-ouia-component-type="PF5/TabButton"
                        data-ouia-safe={true}
                        data-quickstart-id="settings-notifications&messages-tab"
                        id="pf-tab-SETTINGS.CATEGORIES.NOTIFICATION_MESSAGE-719517530"
                        onClick={[Function]}
                        role="tab"
                        type="button"
                      >
                        <span
                          className="pf-v5-c-tabs__item-text"
                        >
                          Notifications & messages
                        </span>
                      </button>
                    </li>
                    <li
                      className="pf-v5-c-tabs__item"
                      role="presentation"
                    >
                      <button
                        aria-controls="pf-tab-section-SETTINGS.CATEGORIES.DASHBOARD-129171981"
                        aria-selected={false}
                        className="pf-v5-c-tabs__link"
                        data-ouia-component-type="PF5/TabButton"
                        data-ouia-safe={true}
                        data-quickstart-id="settings-dashboard-tab"
                        id="pf-tab-SETTINGS.CATEGORIES.DASHBOARD-129171981"
                        onClick={[Function]}
                        role="tab"
                        type="button"
                      >
                        <span
                          className="pf-v5-c-tabs__item-text"
                        >
                          Dashboard
                        </span>
                      </button>
                    </li>
                    <li
                      className="pf-v5-c-tabs__item"
                      role="presentation"
                    >
                      <button
                        aria-controls="pf-tab-section-SETTINGS.CATEGORIES.ADVANCED-843151721"
                        aria-selected={false}
                        className="pf-v5-c-tabs__link"
                        data-ouia-component-type="PF5/TabButton"
                        data-ouia-safe={true}
                        data-quickstart-id="settings-advanced-tab"
                        id="pf-tab-SETTINGS.CATEGORIES.ADVANCED-843151721"
                        onClick={[Function]}
                        role="tab"
                        type="button"
                      >
                        <span
                          className="pf-v5-c-tabs__item-text"
                        >
                          Advanced
                        </span>
                      </button>
                    </li>
                  </ul>
                </div>
              </div>
              <div
                className="pf-v5-c-sidebar__content"
              >
                <form
                  className="pf-v5-c-form settings__content"
                  noValidate={true}
                >
                  <div
                    className="pf-v5-c-form__group"
                  >
                    <div
                      className="pf-v5-c-form__group-label"
                    >
                      <label
                        className="pf-v5-c-form__label"
                      >
                        <span
                          className="pf-v5-c-form__label-text"
                        >
                          <h2
                            className="pf-v5-c-title pf-m-lg"
                            data-ouia-component-id="OUIA-Generated-Title-1"
                            data-ouia-component-type="PF5/Title"
                            data-ouia-safe={true}
                          >
                            Date & Time
                          </h2>
                        </span>
                      </label>
                       
                    </div>
                    <div
                      className="pf-v5-c-form__group-control"
                    >
                      <div
                        className="pf-v5-c-form__helper-text"
                      >
                        <div
                          className="pf-v5-c-helper-text"
                        >
                          <div
                            className="pf-v5-c-helper-text__item"
                          >
                            <span
                              className="pf-v5-c-helper-text__item-text"
                            >
                              
                            </span>
                          </div>
                        </div>
                      </div>
                      <p
                        className=""
                        data-ouia-component-id="OUIA-Generated-Text-1"
                        data-ouia-component-type="PF5/Text"
                        data-ouia-safe={true}
                        data-pf-content={true}
                      >
                        DatetimeControl Component
                      </p>
                    </div>
                  </div>
                  <div
                    className="pf-v5-c-form__group"
                  >
                    <div
                      className="pf-v5-c-form__group-label"
                    >
                      <label
                        className="pf-v5-c-form__label"
                      >
                        <span
                          className="pf-v5-c-form__label-text"
                        >
                          <h2
                            className="pf-v5-c-title pf-m-lg"
                            data-ouia-component-id="OUIA-Generated-Title-2"
                            data-ouia-component-type="PF5/Title"
                            data-ouia-safe={true}
                          >
                            Theme
                          </h2>
                        </span>
                      </label>
                       
                    </div>
                    <div
                      className="pf-v5-c-form__group-control"
                    >
                      <div
                        className="pf-v5-c-form__helper-text"
                      >
                        <div
                          className="pf-v5-c-helper-text"
                        >
                          <div
                            className="pf-v5-c-helper-text__item"
                          >
                            <span
                              className="pf-v5-c-helper-text__item-text"
                            >
                              Set the current theme for web console.
                            </span>
                          </div>
                        </div>
                      </div>
                      <p
                        className=""
                        data-ouia-component-id="OUIA-Generated-Text-2"
                        data-ouia-component-type="PF5/Text"
                        data-ouia-safe={true}
                        data-pf-content={true}
                      >
                        Theme Component
                      </p>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="pf-v5-l-stack__item"
      />
    </div>
  </section>
</div>
`;
